### ISSUE TITLE
hdr tonemapping: selecting trc=bt.1886 instead of srgb breaks the image in dark scenes (saturation, contrast, brightness)

### ISSUE THREAD

**Issue by geextahslex** [mpv-player/mpv#15340]
### mpv Information

```bash
mpv v0.39.0-355-g45c1c586 Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Nov 19 2024 00:07:00
libplacebo version: v7.349.0 (v7.349.0-26-g5788a82-dirty)
FFmpeg version: N-117832-ge9c3698ed
FFmpeg library versions:
   libavcodec      61.25.100
   libavdevice     61.4.100
   libavfilter     10.6.101
   libavformat     61.9.100
   libavutil       59.47.100
   libswresample   5.4.100
   libswscale      8.9.101
```

### Other Information

```text
Windows version: Windows 10 22H2 Build 19045.4170)
GPU model, driver and version: intel HD Graphics 4600, driver: 20.19.15.5171
Source of mpv: https://sourceforge.net/projects/mpv-player-windows/files/
Introduced in version: present since v38.0
```

### Reproduction Steps

use these hdr tonemapping settings and compare a dark scene with and without `vf=format=sig-peak=4.926` to see the difference in saturation
```
tone-mapping=st2094-10
target-peak=100
hdr-compute-peak=no
allow-delayed-peak-detect
```

### Expected Behavior

saturation stays the same

### Actual Behavior

dark scenes get desaturated and look washed out

### Log File
full output
[output full.txt](https://github.com/user-attachments/files/17823500/output.full.txt)
gpu-debug
[output.txt](https://github.com/user-attachments/files/17821363/output.txt)

### Sample Files
almost black and white
![black and white](https://github.com/user-attachments/assets/a54b8e8a-6a0e-409a-8cb2-845546596919)

normal
![normal4545](https://github.com/user-attachments/assets/4a49ca30-771a-459b-b992-134c298936ea)

washed out
![washed out](https://github.com/user-attachments/assets/7e398d5c-8cae-4185-90b3-7c2759881f7c)

normal
![normal3](https://github.com/user-attachments/assets/381d581b-9f00-4475-b816-de01db225185)

### I carefully read all instruction and confirm that I did the following:

- [X] I tested with the latest mpv version to validate that the issue is not already fixed.
- [X] I provided all required information including system and mpv version.
- [X] I produced the log file with the exact same set of files, parameters, and conditions used in "Reproduction Steps", with the addition of `--log-file=output.txt`.
- [X] I produced the log file while the behaviors described in "Actual Behavior" were actively observed.
- [X] I attached the full, untruncated log file.
- [X] I attached the backtrace in the case of a crash.

**Comment by kasper93** [comment]
Likely duplicate of #12707 you can try with `target-contrast=inf`

**Comment by geextahslex** [comment]
This doesn't fix it, this messes up the contrast or black levels

**Comment by geextahslex** [comment]
It seems like I would need to disable this "feature" that desatures the image

> // Avoid raising saturation excessively when raising brightness, and
         // also desaturate when reducing brightness greatly to account for the
         // reduction in gamut volume. 

https://github.com/mpv-player/mpv/issues/12707#issuecomment-1774077987

**Comment by geextahslex** [comment]
My suggestion would be to add an percantage option to this function, so you can turn it on, off or fine-tune it to your liking.

**Comment by stonix015** [comment]
Hey @geextahslex ....
I want to understand more about this: How did you find out that vf=format=sig-peak=4.926 deactivates the feature. What does this do exactly?

Wouldn't it be the same to set target-peak = 1000 ?

So what you want is the Brightness of the HDR picture, but you do NOT want the washed out colors, mhh?

Did you try out the spline tone-mapping instead of st2094 ? It really works out as the best i think. The algorithm makes really sense.

**Comment by haasn** [comment]
https://code.videolan.org/videolan/libplacebo/-/merge_requests/683

**Comment by geextahslex** [comment]
@stonix015 
>I want to understand more about this: How did you find out that vf=format=sig-peak=4.926 deactivates the feature. What does this do exactly?

I found this out by accident, and kasper gave me this line to fix it.

>Wouldn't it be the same to set target-peak = 1000 ?

From my understanding no, because target-peak is kinda your output target and the metadata nits is the input to the pipeline. You can see it at the top in these screenshots.

almost black and white
![black and white](https://github.com/user-attachments/assets/a54b8e8a-6a0e-409a-8cb2-845546596919)

normal
![normal4545](https://github.com/user-attachments/assets/4a49ca30-771a-459b-b992-134c298936ea)

>So what you want is the Brightness of the HDR picture, but you do NOT want the washed out colors, mhh?

Yes, I dont want the desaturation effect.

>Did you try out the spline tone-mapping instead of st2094 ? It really works out as the best i think. The algorithm makes really sense.

At this point I tried out everything, spline, hable, rainhard etc. For me st2094-10 works kinda the best.

**Comment by geextahslex** [comment]
https://github.com/haasn/libplacebo/issues/304

**Comment by geextahslex** [comment]
So after further testing I noticed that this behaviour only occurs when you set the `target-trc` to `srgb/bt.1886` instead of the default `pq`

**Comment by geextahslex** [comment]
So after further testing I ended up using [hdr-toys](https://github.com/natural-harmonia-gropius/hdr-toys) for all of the "color space" stuff (gamut-mapping and transfer function) and mpv only does the tone-mapping with 2094-10. 

On a SDR TV this gives you the best result. No clipped highlights, no crushed blacks, no washed out colors. Also good performance on crappy hardware :)
```
target-prim=bt.2020
target-trc=pq
tone-mapping=st2094-10
gamut-mapping-mode=clip (clip because hdr-toys handles this)
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect

glsl-shader=~~/shaders/hdr-toys/transfer-function/pq_inv.glsl
glsl-shader=~~/shaders/hdr-toys/utils/black_point_compensation.glsl
glsl-shader=~~/shaders/hdr-toys/gamut-mapping/jedypod.glsl
glsl-shader=~~/shaders/hdr-toys/transfer-function/bt1886.glsl
glsl-shader-opts=L_hdr=1000,L_sdr=100
```
![1 dd](https://github.com/user-attachments/assets/d3fc6ed0-b60a-4276-8e38-2e2e1f15c412)
![2 dd](https://github.com/user-attachments/assets/02f3d671-957c-4bdd-accd-10853a6c5400)

**Comment by danbezerra** [comment]
@geextahslex I always get a lot of brightness fluctuation when trying HDR Toys. With this mix do you manage to get stable brightness? I'm also shooting for 100 nits output.

**Comment by geextahslex** [comment]
what do you mean by "brightness fluctuation"? Yes this setup above should work. You can also try this, without hdr-toys

```
target-contrast=1000000
target-trc=srgb
tone-mapping=st2094-10
target-peak=140
hdr-compute-peak=yes
allow-delayed-peak-detect=yes
deband=no
```

**Comment by danbezerra** [comment]
Thanks. It's when the brightness keeps changing within the same scene, and you see it "pumping" in the background of the image.  However, with your solution you leave MPV to take care of the tone-mapping. This can be a good choice. I will test it later.

**Comment by geextahslex** [comment]
Okay, so with hdr-toys it works nice.

**Comment by geextahslex** [comment]
ok so what I can add to this is that the lower the `PQ-Y Average` goes like 10-5%, the more you have to add `saturation` manually, like about `+50` to counter this effect

**Comment by Headcrabed** [comment]
@geextahslex Btw have you actually tested if https://code.videolan.org/videolan/libplacebo/-/merge_requests/683 helps?

**Comment by geextahslex** [comment]
@Headcrabed yes I was responsible to stop this patch as it made it look even worse
https://github.com/mpv-player/mpv/discussions/15375

**Comment by geextahslex** [comment]
New screenshots on the latest mpv version
broken, desaturated, black&white (default mpv)
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/a1f8f457-8a2a-4c7d-ae6c-f687fa2c7a24" />


how it should look like, with colors (workaround profile)
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/3d09058e-c05f-400a-b59b-dee4d579ea70" />

**Comment by kasper93** [comment]
If anyone is interested in testing new tone mapping (and gamut mapping) changes. I've been focusing on the clipping and saturation issues we've had at certain tone mapping ratios.

Windows builds here: https://github.com/kasper93/mpv/actions/runs/30770795463 (or https://nightly.link/kasper93/mpv/actions/runs/30770795463)

What I'm most interested in is where it still fails: a sample (or at least the title + timestamp) and your exact target parameters (screen shot of stats page), so I can reproduce and tune it. The current version is early, so don't hesitate to call things out. Comparisons to old build are welcome too, if you think it was better.

What changed, in short:

- Dynamic tone mapping target. Instead of compressing everything straight down to the display peak, it lowers average brightness a bit to create headroom, based on the measured FALL. Mapping something like 10000 -> 50 nits statically just isn't feasible, and this gives a much more contrasty and pleasing image in those cases. Bright, low-range content and mild HDR->HDR mappings (e.g. 1000 -> 600) are mostly unaffected. Note this relies on peak detection, so keep --hdr-compute-peak enabled (it is by default).
- The tone mapping itself now preserves saturation properly. Previously saturation was reapplied after the fact and, depending on the target, ended up either oversaturated or washed out. That should hopefully be fixed now. Dark scenes on low-contrast targets should no longer desaturate either.
- One change that may bite on some samples: colors above the declared mastering display peak are now softly clipped toward it. This fixes titles like MMFR, which is full of artifacts in colors above the mastering peak. It's on by default. If you suspect it's misbehaving on something, you can toggle it with --libplacebo-opts=mastering_clip=0 (or =1) and compare.
- 
EDIT: I found an issue, which made image too dark. I will fix tommorow, you can skip testing this version for now, as you probably notice mostly this on bright scenes.

**Comment by Jules-A** [comment]
> If anyone is interested in testing new tone mapping (and gamut mapping) changes. I've been focusing on the clipping and saturation issues we've had at certain tone mapping ratios.

Uggh... The timing... I literally spent over an hour yesterday trying to fix MPV's awful HDR with shaders... I think it currently fails so bad because my monitor is above 1000 nits and tonemapping/gamut-mapping only seems to be designed to go to 1000. Biggest issue is the image is completely desaturated followed by highlights looking miserable and then the overdarkened scenes are probably the least of the major concerns.
I'll test it once you've made some more fixes and I get more time...

**Comment by geextahslex** [comment]
@kasper93 Thats awesome news, finally something is happening here. I will gladly test it out and give you feedback.
I don't know if you have seen my [feature-request](https://github.com/mpv-player/mpv/issues/18331) and if this feature could be implemented now, to give more direct control over whats happening. So you don't have to rely on lua or glsl that much, because at the moment I have to boost manually `gamma+10` in the darkest scenes like `AVG-PQ` 10% and below. 
In short, I feel like `hdr-compute-peak` is an awesome light weight system, but it gives you only `on/yes` or `off/no` option, an ability to correct something would be great. 
I wait for the fixed version :)

@Jules-A I have also spend the last 2 days on testing pure mpv (without glsl and lua) to have a fallback or a more performant option but nothing gave pleasent results. Lets hope this new update helps here :)

**Comment by kasper93** [comment]
Since yesterday's build I've done some more testing and resolved the most pressing issues. I've also added "bloom restoration" and perceptual desaturation of very bright saturated areas. Please give it a try, any feedback is welcome.

Windows build: https://github.com/kasper93/mpv/actions/runs/30857763541 (or https://nightly.link/kasper93/mpv/actions/runs/30857763541)
PR if you want to build yourself: https://code.videolan.org/videolan/libplacebo/-/merge_requests/875
Some comparisons of `target-peak=80 target-contrast=inf` between master and new: https://slow.pics/c/jwAydqri

Note that I've added some experimental tweaks and I'm interested in your feedback, especially on the slight desaturation, which is mostly visible on color-ramp test patterns.

Any feedback is welcome, the more vocal you are the better we can tune this.

**Comment by Headcrabed** [comment]
My config:
```
vo=gpu-next
gpu-api=vulkan
hwdec=vaapi
treat-srgb-as-power22=both
```
Using Fedora KDE 44 Wayland.
```
[   0.057][v][vo/gpu-next/wayland] target: min_luma=0.153600, max_luma=140.000000, max_cll=140.000000, max_fall=140.000000
[   0.057][v][vo/gpu-next/wayland]         raw prims: red.x=0.688950,   red.y=0.306778,
[   0.057][v][vo/gpu-next/wayland]                    green.x=0.263443, green.y=0.662994,
[   0.057][v][vo/gpu-next/wayland]                    blue.x=0.146042,  blue.y=0.062729,
[   0.057][v][vo/gpu-next/wayland]                    white.x=0.308758, white.y=0.322312
```


(Left side is original upstream code, right side is PR version)
`tone-mapping=st2094-40`:

<img width="2560" height="722" alt="Image" src="https://github.com/user-attachments/assets/9dfd36a8-1b23-4fb6-817c-8dea459918a4" />

<img width="2560" height="719" alt="Image" src="https://github.com/user-attachments/assets/5a4fcb12-5fb5-467b-81c1-0378ac68fc36" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/e547a6d4-50ed-4c78-a7cc-61586dc9b594" />

<img width="2560" height="721" alt="Image" src="https://github.com/user-attachments/assets/cad85a10-7c8c-46ca-abc1-2d3023d878e1" />



`tone-mapping=st2094-10`:

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/7c3ae4a9-0417-4325-8161-63a2b7301eac" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/7227984e-8f39-4a4d-8dbc-1adc9dce6ce5" />

<img width="2560" height="722" alt="Image" src="https://github.com/user-attachments/assets/f2187b21-09f1-4488-9836-4e593cc998cc" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/006a5b29-7d4f-4892-be04-d587b024b331" />



`tone-mapping=bt.2446a`:

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/a2dd96e6-db1e-42c5-8b0f-0cc16711efef" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/339c9e68-665a-443c-9de4-6e4b83243ebc" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/81c22638-9bce-484a-9f24-80cdd93809d8" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/c7d15bfc-3980-4ce1-b625-f41f958b3657" />



Seems brightness in PR branch is always lower than previous version. And in dark scenes, it's basically too dark to see anything in PR branch.

**Comment by Jules-A** [comment]
Just a quick test watching HDR with HDR output:
It's still extremely undersaturated, maybe a slight improvement but I just went off memory so can't say for sure.
Overdarkened scenes seems fixed, maybe slightly too dark but that's probably just from the undersaturation
Highlights seems ever so slightly better but still pretty miserable.

One issue I noticed is I can't seem to get real blacks in scenes that I used to that have any bright elements occupying the scene. The brightness seems to radiate throughout the whole scene when it shouldn't be.

**Comment by kasper93** [comment]
> One issue I noticed is I can't seem to get real blacks in scenes that I used to that have any bright elements occupying the scene. The brightness seems to radiate throughout the whole scene when it shouldn't be.

Sample?

**Comment by geextahslex** [comment]
I can confirm the "it just makes everything darker" behaviour
my full profile is available here https://github.com/mpv-player/mpv/discussions/14757

samples
[dark sample2.zip](https://github.com/user-attachments/files/30598960/dark.sample2.zip)
[HDR Snow.zip](https://github.com/user-attachments/files/30707429/HDR.Snow.zip)

Windows 10 x64 22H2
```
hwdec=no
vo=gpu-next
dither-depth=8
dither=fruit

target-colorspace-hint=no
gamut-mapping-mode=clip
vf=format:dolbyvision=no:hdr10plus=no
tone-mapping=st2094-10
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect=no
target-prim=bt.2020
target-trc=pq

#GLSL SHADER
glsl-shader=~~/shaders/pq_inv.glsl
glsl-shader-opts=L_hdr=1000,L_sdr=100
glsl-shader=~~/shaders/black_point_compensation.glsl
glsl-shader=~~/shaders/jedypod.glsl
glsl-shader=~~/shaders/bt1886.glsl
```

old mpv
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/a001744e-5a96-4912-be82-f7006aadb1f5" />
current testing
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/50a4d5ea-2c95-4f06-9fc4-efed1a2d82dc" />

old mpv
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/c933b3f5-3e1f-4413-a9aa-278dbfc591fa" />
current testing
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/93b97aec-df43-4553-a286-4805c6b5a17e" />

**Comment by Jules-A** [comment]
> Sample?

Ah whoops, I was mistaken, I verified with MadVR, that scene didn't actually have real blacks because the glow on the smaller bright objects were meant to light the whole scene. It was just the increased brightness on the object that caused it to be slightly brighter throughout.

Also, I did some follow up blind testing and I couldn't find any differences in saturation outside of changes from brightness.

I did notice some other issues, like the scene from this sample (LG video downloaded from YT): https://filebin.net/dphrb70fzii20our
which had HDR10+ attributes added and highlights are significantly worse:
Master:
<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/3ce36e2d-338d-4386-8218-789f867e31e6" />
PR:
<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/4e318fc2-c5eb-4c0d-a4ef-08827df49be2" />

Note: The image looks nothing like the screenshots, for some reason I can't get MPV screenshots to output correctly in HDR.

**Comment by kasper93** [comment]
> I can confirm the "it just makes everything darker" behaviour

This is what you requested in the top post, your comparisons are clear, washed out vs normal. This is expected that we shift the target nits (DTN) to compensate for average luminance.

**Comment by geextahslex** [comment]
@kasper93 
No, I said "the image gets almost black and white in a dark scene" so only the dark scenes were broken, the bright scenes were fine.
Now both (bright and dark scenes) get darker by the same amount. 

Look at this sample with this config
The normal and test mpv. The first scene has the same brightness with both (the colors are much better now) but in the 2nd scene you get this huge fall in brightness with the test version, the old one keeps the floor much higher.
[sample luma3.zip](https://github.com/user-attachments/files/30720340/sample.luma3.zip)
```
hwdec=no
vo=gpu-next
dither-depth=8
dither=fruit
target-colorspace-hint=no
vf=format:dolbyvision=no:hdr10plus=no
tone-mapping=st2094-10
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect=no
hdr-peak-percentile=99.9
hdr-peak-decay-rate=200
hdr-scene-threshold-low=2.0
hdr-scene-threshold-high=6.0
```
red row is the test mpv, the blue is the normal mpv. It would be great if we could get the green combination Luma wise
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/b4f40055-9f39-40b2-bf8a-901fbc8775e1" />

In short: 
keep both samples at the same brightness but keep colors/saturation (as you have already fixed that)
[dark sample2.zip](https://github.com/user-attachments/files/30598960/dark.sample2.zip)

**Comment by kasper93** [comment]
I've fixed BPC in previous build. No other significant changes, but the BPC fix alone will affect dark scenes on non-inf contrast.

Windows build: https://github.com/kasper93/mpv/actions/runs/30962926900 (or https://nightly.link/kasper93/mpv/actions/runs/30962926900)
PR if you want to build yourself: https://code.videolan.org/videolan/libplacebo/-/merge_requests/875

> red row is the test mpv, the blue is the normal mpv. It would be great if we could get the green combination Luma wise

I don't know, might be fixed with new build. Try it.

>  It's still extremely undersaturated, maybe a slight improvement but I just went off memory so can't say for sure.

Cannot reproduce. Need samples. For me saturation is fixed, for example here is `target-contrast=1000`

<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/a2b8e364-e1fe-4bf9-ade3-97c6824f3654" />
<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/35f5c8a6-4e7a-4f84-a771-fb89d1557570" />

**Comment by geextahslex** [comment]
@kasper93 
That looks awesome. I dont know what happend but here with the torch flickering in the background you get huge lagging and flickering, like 1 FPS

```
hwdec=no
vo=gpu-next
dither-depth=8
dither=fruit
deband=no
target-colorspace-hint=no
vf=format:dolbyvision=no:hdr10plus=no
tone-mapping=st2094-10
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect=no
```

[H lagg.zip](https://github.com/user-attachments/files/30725202/H.lagg.zip)
P.S. it's not only this sample but in general all files, but this sample is veeery slow.

**Comment by kasper93** [comment]
> I dont know what happend but here with the torch flickering in the background you get huge lagging and flickering, like 1 FPS

Fixed: https://github.com/kasper93/mpv/actions/runs/30969004465 (or https://nightly.link/kasper93/mpv/actions/runs/30969004465)

Some shader variables were not dynamic and caused recompilation.

**Comment by Jules-A** [comment]
> Cannot reproduce. Need samples. For me saturation is fixed, for example here is `target-contrast=1000`

It happens on every HDR video I've tried, even the one I sent you previously though I guess I could find one where it's more obvious.

**Comment by Headcrabed** [comment]
Test result on newest PR branch:

Dark scenes look much better now, however the last test video is desaturated and unchanged since 1st version.

`tone-mapping=st2094-40`:

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/2c67caeb-4836-4977-9f31-e3413ae73c4e" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/d7a404d7-4267-4a4c-9007-a781fdcb0add" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/1c450756-f1ed-4781-a74a-c15350a67ec4" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/1f0d7aac-d9c7-4e6a-b1a6-3300a3250ee3" />

`tone-mapping=st2094-10`:

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/eb277b80-41c3-4849-a4aa-7843de289182" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/029ab77c-a80e-471a-8680-a94b7e3c297e" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/f94b4176-6b2c-448c-bbeb-49cff8cccbec" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/f47fd76d-4706-4562-b6ef-f7ed5d3d6d24" />

`tone-mapping=bt.2446a`:

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/256f102a-c2bc-4c05-b079-ee1c1218f2ad" />

<img width="2560" height="719" alt="Image" src="https://github.com/user-attachments/assets/a9a05d04-8790-4635-a771-97cd7a56feb1" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/60ef31e9-3cdc-4b0f-8813-6bcdc9f7303e" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/5955459d-8dd1-461d-a341-5e3dc700d324" />

**Comment by geextahslex** [comment]
The performance is fixed while normal watching, it is even worse with `tone-mapping-visualize=yes`
And there is still much flickering in this "torch" scene even with "relaxed hdr-peak timings"
```
hdr-peak-percentile=99.9
hdr-peak-decay-rate=200
hdr-scene-threshold-low=2.0
hdr-scene-threshold-high=6.0
```
But it's slowly getting there, awesome :)

**Comment by kasper93** [comment]
> however the last test video is desaturated and unchanged since 1st version.

This is expected. I will try to tune this, but generally bright light sources are "blinding", and you don't get to see much color. Consider those two examples. Sure the sky sample is overtuned, but those light shafts bright and natural lighting, else it looks too flat.

<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/41ad939f-372a-4848-ba32-2b647a45b989" />
<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/4eb2d40c-a3e3-47e0-9534-2fe21130674e" />

<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/ff63dc4f-2fb0-4f62-9d95-8ab0d6c2d87d" />
<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/1d99b7ce-7e0e-4d0e-a702-1a773391e4dc" />

**Comment by geextahslex** [comment]
I dont know how important that is at the moment but only bt.1886 gives you artifact free colors, the default gamma 2.2 (and hlg) has issues

<img width="1920" height="996" alt="Image" src="https://github.com/user-attachments/assets/37804b31-79dd-4f74-9643-fd4a1b30b17a" />
<img width="1920" height="1002" alt="Image" src="https://github.com/user-attachments/assets/fb21f6e8-2342-4ffe-bde2-c38a7c8d23ab" />

[Daredevil color test.zip](https://github.com/user-attachments/files/30748170/Daredevil.color.test.zip)

**Comment by kasper93** [comment]
> I dont know how important that is at the moment but only bt.1886 gives you artifact free colors, the default gamma 2.2 (and hlg) has issues

What artifacts you mean? You are seeing as video is encoded. bt.1886 should never be used on output. It basically is clipping your output. Adjust your `--target-contrast` for a value that matches your expectations and display.

**Comment by geextahslex** [comment]
It's difficult to make them visible. Here I boosted brightness in photoshop. I mean something like banding when colors go from darker to brighter but its pixelated. bt.1886 looks smooth and gamma 2.2 has these weird chunks/wholes in the transition 
gamma 1.8 makes this less obvious/visible, I just wanted to point that out, as we are now in the "testing phase"

<img width="1920" height="1002" alt="Image" src="https://github.com/user-attachments/assets/8cc76d40-5a6e-4e44-9dbc-740c34aa66eb" />
<img width="1920" height="1002" alt="Image" src="https://github.com/user-attachments/assets/9000e908-c204-4605-854f-60fca9b494c1" />

**Comment by kasper93** [comment]
No new build today, but you can pause testing on previous builds. I have been cooking something, to fix some corner cases, will share later.

**Comment by Headcrabed** [comment]
> It's difficult to make them visible. Here I boosted brightness in photoshop. I mean something like banding when colors go from darker to brighter but its pixelated. bt.1886 looks smooth and gamma 2.2 has these weird chunks/wholes in the transition gamma 1.8 makes this less obvious/visible, I just wanted to point that out, as we are now in the "testing phase"


I think using what gamma more depending on your OS configuration. What’s your enable status of ACM/HDR?

**Comment by kasper93** [comment]
I wanted to share build, but got side tracked into tuning this one. There is a balance from preserving hue and slight bend of it to make it more pleasant to look and punchy. If you would like to look at https://slow.pics/c/P1L89tqi and see which variant looks best for you. The most interesting samples are: bvs, madmax, madmax2, demo1, demo4, gotg, amaze.

Tell me which one (`fs...`) you like best :)

**Comment by Jules-A** [comment]
> Tell me which one (`fs...`) you like best :)

None of them really looked correct. If you account for the sample with the spear it needs around fs80 until it looks correct but I don't know what it really is supposed to look like, what sample is that? Otherwise for the rest I'd say somewhere between fs40 and fs50, leaning almost entirely towards fs50.

**Comment by geextahslex** [comment]
Isn't it possible to make this to new a variable, so anyone can adjust it to their equipment and taste?
We all know that HDR to SDR is not "all size fits all" so more knobs to tune would be nice.
It's hard to judge on not your own samples, because you don't know which bothered you while watching and which were "fine"
fs50 seems like the safe middleground value, but having control over it in the future with a variable would be awesome.
A lower value looks dull/clamped, a higher value looks better but could blow out?

You could include these samples for more highlight stuff
[HDR samples.zip](https://github.com/user-attachments/files/30823035/HDR.samples.zip)

**Comment by roberth1990** [comment]
There still need to have a default regardless, can't expect everyone to having to test and adjust it.

**Comment by Headcrabed** [comment]
> I wanted to share build, but got side tracked into tuning this one. There is a balance from preserving hue and slight bend of it to make it more pleasant to look and punchy. If you would like to look at https://slow.pics/c/P1L89tqi and see which variant looks best for you. The most interesting samples are: bvs, madmax, madmax2, demo1, demo4, gotg, amaze.
> 
> Tell me which one (`fs...`) you like best :)

I like fs040 and fs020. Can we get something like fs030/35?

**Comment by kasper93** [comment]
> None of them really looked correct.

Tough. I don't have better version for you.

> If you account for the sample with the spear it needs around fs80 until it looks correct but I don't know what it really is supposed to look like, what sample is that?

BvS is Batman v Superman. Either way, we shouldn't optimize for one overcooked movie and scene. For me fs70 is rather good for this scene, but it trades of too much on others.

In general on this test set and other samples fs50, seems to be mostly balanced, for me.

> Otherwise for the rest I'd say somewhere between fs40 and fs50, leaning almost entirely towards fs50.

fs40 is also an option, but the difference here is already so small, that there is no point in splitting the hair in two. We only see a difference, because we directly compare it.

> Isn't it possible to make this to new a variable, so anyone can adjust it to their equipment and taste?

99.9% users will never adjust that, and even you will play around with this and land on some default value that we agree on. And we still need default value. It's not like you will be adjusting this strength for each movie you watch.

> We all know that HDR to SDR is not "all size fits all" so more knobs to tune would be nice.

No they are not. We may have one or two options, but the minor things like that should be tuned and left alone. 99.99999% of users want to watch movie, not develop algorithm before watching. 

And yes, I plan do add this single option, but like I said, I think you overestimate user ability to be able to adjust things like that. Remember that on vast majority of scenes this strength does nothing.

> A lower value looks dull/clamped, a higher value looks better but could blow out?

Not really blow out. The higher values are desaturated too much, because we bend hue too much. It makes sunset scene for example lose it's red sunset look, if set too much. But in general, it's still pleasing look.

> There still need to have a default regardless, can't expect everyone to having to test and adjust it.

Correct.

> I like fs040 and fs020. Can we get something like fs030/35?

fs40 is fine by me too. But below that it's bit harsh, on some samples.

**Comment by Headcrabed** [comment]
For me, fs50 often get too washed out. For me I think something between fs25-fs40 is OK.

**Comment by Jules-A** [comment]
> BvS is Batman v Superman. Either way, we shouldn't optimize for one overcooked movie and scene. For me fs70 is rather good for this scene, but it trades of too much on others.

Is it official source or issue with source content? The reason why I was leaning towards 50 was to account for the extremes like the fire and lightning that were clipping similar to the spear. If I were to ignore them then 20 looks the best since 40+ the image looks like the scene has been slightly overbrighted. If there was a 30 option then that would likely be better.
BUT... Accounting for worse case scenes is way more important since you can likely work around the brightness/saturation issues with shaders but trying to fix it after it's already been clipped is near impossible... (it would be easier if we had an option to completely disable and override with shaders though).

> fs40 is also an option, but the difference here is already so small, that there is no point in splitting the hair in two. We only see a difference, because we directly compare it.

Not from what I'm seeing, going from 40->50 is completely stopping any clipping in some scenes at the sacrifice of a very noticeable amount of saturation loss that can completely change stuff like sky colour...

> No they are not. We may have one or two options, but the minor things like that should be tuned and left alone. 99.99999% of users want to watch movie, not develop algorithm before watching.

Eh, I see hundreds of threads with people trying to get tonemapping/gamut mapping working in MPV and I myself have spent an insane amount of time trying to get HDR usable in MPV and while I did manage to get it kind of acceptable with shaders, it's still far inferior to MadVR with the exception of not having the annoying glowing subtitle issue.
Also yeah, this thread is technically about HDR->SDR but the underlying issue is the same.

**Comment by kasper93** [comment]
New changes are in. This includes everything from recent days, in both tone and gamut mapping, resolving many corner cases I was chasing. Including a physically-motivated film look with a new option to control its strength: `--libplacebo-opts=film_strength=0.5`

Quick explanation of the film response. Where the tone curve compresses highlights, the color channels converge the way print film does, so bright emissive gradients (fire, neon, sky glow) roll toward white instead of flattening into a single color. The tradeoff is hue/saturation fidelity of very saturated colors. Full film response reads as punchy highlights but bends e.g. deep orange toward yellow. 0 gives you strictly chromaticity-preserving tone mapping (exact colors, flatter highlights), 1 the full film response, and the default 0.5 is the balance calibrated on a wide sample set. 

Note that the scaling works differently than in the images shared previously, so don't use the values from before.

Also if you want to have the more clipped look, you can lower `hdr-peak-percentile`

Windows build: https://github.com/kasper93/mpv/actions/runs/31204403351 (or https://nightly.link/kasper93/mpv/actions/runs/31204403351)
PR if you want to build yourself: https://code.videolan.org/videolan/libplacebo/-/merge_requests/875

**Comment by geextahslex** [comment]
Looks really good :)
`target-peak` doesn't seem to have any effect at the moment.
is display meant to be always at 80? `target-peak=100` (this is my setting)

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/6d3ee9d8-3080-44f3-8952-0d1fbdcbdc0f" />

**Comment by Headcrabed** [comment]
> New changes are in. This includes everything from recent days, in both tone and gamut mapping, resolving many corner cases I was chasing. Including a physically-motivated film look with a new option to control its strength: `--libplacebo-opts=film_strength=0.5`

I wonder if we could have this option as a dedicated mpv option, rather than using `libplacebo-opts`? Most other options already done this.

**Comment by kasper93** [comment]
> `target-peak` doesn't seem to have any effect at the moment.
> is display meant to be always at 80? `target-peak=100` (this is my setting)

Try setting `--hdr-reference-white=100`. This is not currently related to changes I work on. Maybe something is wrong with those override options. 80 nits is requested by Windows. I will look later. Override works for me, so not sure I will be able to reproduce.

**Comment by kasper93** [comment]
> > New changes are in. This includes everything from recent days, in both tone and gamut mapping, resolving many corner cases I was chasing. Including a physically-motivated film look with a new option to control its strength: `--libplacebo-opts=film_strength=0.5`
> 
> I wonder if we could have this option as a dedicated mpv option, rather than using `libplacebo-opts`? Most other options already done this.

Later. It's not related to current changes.

**Comment by geextahslex** [comment]
`hdr-reference-white=100` that worked, its now as before 0.1/100 cd/m

**Comment by kasper93** [comment]
Also, for better mid tones brightness stability you can use `--tone-mapping=bt.2390` , but ymmv.

**Comment by kasper93** [comment]
New version. I've changes desat to LMS, such that we preserve hue. `film_strength` will work different now, but overall it's improvement. Current value is 0.3. In practice, higher values give strongly compressed highlights (fire, explosions, bright emissive gradients) more punch. They whiten towards their peak the way they would on a brighter display, instead of  flattening into a dim uniformly-colored patch. Lower values keep such highlights fully saturated at the cost of that intensity.

EDIT: I made the `film_strength` accept values up to 2, which will give you even more "punch". Value of 1.0 is nominal, above that you may see clipping, but I know people prefer that.

Windows build: https://github.com/kasper93/mpv/actions/runs/31319028305 (or https://nightly.link/kasper93/mpv/actions/runs/31319028305)
PR if you want to build yourself: https://code.videolan.org/videolan/libplacebo/-/merge_requests/875

**Comment by Jules-A** [comment]
> New version. I've changes desat to LMS, such that we preserve hue. 

I didn't test the last version so comparison is vs the build before.

Highlights are better and saturation is very slightly better but unfortunately it's nowhere near enough and not even setting it to 0 is enough :/ Setting it to 0 and increasing global saturation by 2 isn't exactly awful but it essentially undoes all the work on getting highlights better though and makes them miserable once again...

**Comment by kasper93** [comment]
> Highlights are better and saturation is very slightly better but unfortunately it's nowhere near enough and not even setting it to 0 is enough :/ 

I don't know what do you expect. For maximum saturation you can try `--libplacebo-opts=film_strength=0 --gamut-mapping-mode=clip`, this keeps all out of gamut colors as most saturated for your current output. If that's not enough, it's not tone mapper job to make it more saturated and otherwise something else is need changes in your setup.

**Comment by Jules-A** [comment]
> I don't know what do you expect. For maximum saturation you can try `--libplacebo-opts=film_strength=0 --gamut-mapping-mode=clip`, this keeps all out of gamut colors as most saturated for your current output. If that's not enough, it's not tone mapper job to make it more saturated and otherwise something else is need changes in your setup.

Even with clip it's still pretty bad. I'm not so sure it's the tonemapping either and probably something broken with gamut mapping. That said, tone-mapping is so weird now... I see it tonemapping from ~3000 to my monitor's 1083 nits in some scenes on 1000 nit mastered content and then sometimes it says ~300 mapped to ~150.

**Comment by kasper93** [comment]
> Even with clip it's still pretty bad.

Clip is very basic, it does nothing, but drop away any color outside your target.

> I'm not so sure it's the tonemapping either and probably something broken with gamut mapping.

Works on my machine.

> That said, tone-mapping is so weird now... I see it tonemapping from ~3000 to my monitor's 1083 nits in some scenes on 1000 nit mastered content and then sometimes it says ~300 mapped to ~150.

What's weird about that?

**Comment by Headcrabed** [comment]
Tested on newest branch:

`tone-mapping=st2094-40`

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/206faa51-a764-4fd2-807e-5810966dbab5" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/c9dd9a9b-25f9-4c19-bfda-0efac588eefd" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/2132d490-cb4f-49a3-82ee-3c8f58c806a2" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/d1ac9b60-8de6-4f20-87d2-a9b41de183d5" />

In my opinion, current implementation is good enough.

**Comment by pinkflames** [comment]
I think it's specifically st2094-40 that's largely unaffected, bt.2446a on the other hand does respond to changes to film_strength.

I ended up settling on `libplacebo-opts=film_strength=1/3` just because I wanted something a bit more dramatic than 0.3 and 1/3 is conveniently 0.(3).

Best HDR experienced I have ever had but not sure how much it's actually due to the tuning rather than whatever other changes that PR did. As a perhaps relevant side note, letting mpv/libplacebo use the information provided by kwin_wayland now looks better than setting `target-contrast=inf`.

**Comment by netExtra** [comment]
Still issues with SVP  not working with @kasper93's build and since other IPC socket applications do work I think it might be something to do with VapourSynth. Maybe it's missing from his build?

**Comment by geextahslex** [comment]
Can we get a version with working real time AVG-PQ values/monitoring with `hdr-compute-peak=yes`? Now it just shows one static value.
I need AVG-PQ for my lua script to work properly.

**Comment by Headcrabed** [comment]
`tone-mapping=bt.2446a`:

<img width="2560" height="721" alt="Image" src="https://github.com/user-attachments/assets/9cc64732-674c-4ebb-be71-4555ca7bd5f5" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/65897249-6804-4933-b561-c4f15ee783de" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/86ca7685-7a9c-44c3-935d-f7e87a479c7f" />

<img width="2560" height="720" alt="Image" src="https://github.com/user-attachments/assets/c0c866b2-20d1-455a-ac96-94c20d7f3586" />

@pinkflames I think that's just that `bt.2446a` desaturates too much, it desaturates more even without this PR. For contrast, if you provide Kwin with proper ICC which includes actual contrast value, current mpv could get that value from Kwin and use it to get best result.

**Comment by Headcrabed** [comment]
> Can we get a version with working real time AVG-PQ values/monitoring with `hdr-compute-peak=yes`? Now it just shows one static value. I need AVG-PQ for my lua script to work properly.

@kasper93 Seems current PR branch break this. It just fixed at one value.

**Comment by pinkflames** [comment]
@Headcrabed: if you're using Linux 7.1 series and, I think, an older AMD GPU, it's possible you're affected by a driver bug which caused BT.2020 YCbCr to have slightly wrong luma values. The fix will arrive in Linux 7.2.0 due out in some hours probably. The fixes are marked for backporting but no idea, if a fixed 7.1 kernel is out yet. And, if you're using an even older kernel, the HDR there was even more broken... And last time I tried the Xe iGPU the colors were even more off but hopefully they have improved the driver since then. In other words, HDR is amazing when it works but it's sure not the easiest thing in the world to get right.

**Comment by Headcrabed** [comment]
@pinkflames I'm using 7800XT+kernel 7.1.8. For BT.2020 YCbCr conversion, it's not affected on my side, since I'm using RGB full-range output, and the conversion for video yuv-rgb is done by libplacebo so this part also not affected.

**Comment by geextahslex** [comment]
When will this get pushed to the main release build?

**Comment by Jules-A** [comment]
> > That said, tone-mapping is so weird now... I see it tonemapping from ~3000 to my monitor's 1083 nits in some scenes on 1000 nit mastered content and then sometimes it says ~300 mapped to ~150.
> 
> What's weird about that?

I don't understand why it's going so much higher than the source? Also I don't understand why it would map from 300 to 150 instead of mapping from my monitor's 1083 to 150?

**Comment by geextahslex** [comment]
Any news on whats happening with this at the moment? Do we have to wait on a merge on libplacebo?
https://code.videolan.org/videolan/libplacebo/-/merge_requests/875

**Comment by Headcrabed** [comment]
@geextahslex https://code.videolan.org/videolan/libplacebo/-/merge_requests/875#note_579260

**Comment by geextahslex** [comment]
I have found multiple samples where even with relaxed `hdr-scene-threshold` timings you get these sudden exposure "jumps/adjustments"
(Sample 1) at second 2 (when she turns her head from right to left, before the car scene) this happens also without any config
But there is no brightness change in this scene, this also doesnt happen on the normal release mpv

Thats a bit weird because overall the image is actually very stable, so this seems like a "bug" or something.

multiple samples with this behavior
[sample.zip](https://github.com/user-attachments/files/32170255/sample.zip)
[sample 2.zip](https://github.com/user-attachments/files/32170486/sample.2.zip)
[sample 3.zip](https://github.com/user-attachments/files/32170608/sample.3.zip)
[sample 4.zip](https://github.com/user-attachments/files/32171265/sample.4.zip)

**Comment by kasper93** [comment]
This file's metadata is broken, the luminance values jump on an almost static scene. At the I-frame around 2.08s (`lioness.s03e02.mkv`) the HDR10+ metadata starts a new "scene" mid-shot and its stats step from maxscl 90.9 nits / avg 1.7 nits to maxscl 51.3 nits / avg 2.3 nits, while the actual content is continuous, measured maxRGB peak stays at ~92 nits on both sides and the average rises smoothly. The second metadata scene claims a 51 nit maximum over 92 nit pixels, and tone mapping follows the metadata step. You can work around this by stripping HDR10+ metadata with `--vf=format=hdr10plus=no`. However, you need latest master with https://github.com/mpv-player/mpv/commit/0b7ed670f7c353dd3dd4f8ae0fc788a181a15aa6, because I forgot to merge it before.

EDIT: It also is compounding on top of elevated black level that we have by default now. You can use `--target-contrast=1000` to minimize it. I have fixes locally for that, but that's another story.

**Comment by geextahslex** [comment]
Okay thats good news, this complete episode was bouncing up and down. Watching 2 seasons of Reacher, i didnt have this issue, thats why I was wondering what is causing this.
I thought that using `hdr-compute-peak=yes` ignores any metadata and messures its own data. I thought thats the point of `hdr-compute-peak=yes`
I am already using `vf=format:dolbyvision=no:hdr10plus=no` thats my default config

So are all these HDR changes now in the master release?
I can not compile, so if there are any new binaries I will continue testing.

**Comment by kasper93** [comment]
> I thought that using `hdr-compute-peak=yes` ignores any metadata and messures its own data. I thought thats the point of `hdr-compute-peak=yes`

No, metadata is still authoritative. 

> I am already using `vf=format:dolbyvision=no:hdr10plus=no` thats my default config

This is to ignore metadata and use peak detection.

> So are all these HDR changes now in the master release? I can not compile, so if there are any new binaries I will continue testing.

No, changes are not in master.

**Comment by geextahslex** [comment]
Could someone compile a version for windows with this fix?

**Comment by Dkfc123** [comment]
> Could someone compile a version for windows with this fix?

https://github.com/Dkfc123/mpv/actions/runs/34814005841/artifacts/10335924422

**Comment by geextahslex** [comment]
@Dkfc123 Awesome, thank you :)

This solves my issue from [before](https://github.com/mpv-player/mpv/issues/15340#issuecomment-5655936840) and AVG-PQ is also working.
The last thing that I notice is the performance with `tone-mapping-visualize=yes` I don't know if this is meant to be that heavy but not really relevant for normal users.
